File: src/main/kotlin/br/com/caethas/atto/atto/modules/membro/entity/MembroEntity.kt
package br.com.caethas.atto.atto.modules.membro.entity

import br.com.caethas.atto.atto.modules.membro.dto.MembroDto
import br.com.caethas.atto.atto.modules.membro.enums.StatusAssociado
import br.com.caethas.atto.atto.shared.base.BaseEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.Table
import java.util.UUID

@Entity
@Table(name = "membros")
class MembroEntity(
    @Column(name = "membros_id", nullable = false, updatable = false, unique = true)
    var membrosId: UUID? = null, // ID externo vindo do app

    @Column(name = "nome_membro", nullable = false, updatable = true)
    var nomeMembro: String = "",

    @Column(name = "email_membro", nullable = true, updatable = true)
    var emailMembro: String? = null,

    @Column(name = "contato_membro", nullable = true, updatable = true)
    var contatoMembro: String? = null,

    @Column(name = "endereco_membro", nullable = true, updatable = true)
    var enderecoMembro: String? = null,

    @Enumerated(EnumType.ORDINAL)
    @Column(name = "status_associado", nullable = false, updatable = true)
    var statusAssociado: StatusAssociado = StatusAssociado.ATIVO,
) : BaseEntity() {
    fun toDto() = MembroDto(
        id = this.id,
        membrosId = this.membrosId,
        nomeMembro = this.nomeMembro,
        emailMembro = this.emailMembro,
        contatoMembro = this.contatoMembro,
        enderecoMembro = this.enderecoMembro,
        statusAssociado = this.statusAssociado.value,
        criadoEm = this.criadoEm,
        atualizadoEm = this.atualizadoEm,
        deletadoEm = this.deletadoEm
    )
}

File: lib/src/modules/membro/entity/membro_entity.dart
import 'package:atto_api/src/modules/membro/dto/membro_dto.dart';
import 'package:atto_api/src/modules/membro/enums/status_associado.dart';
import 'package:atto_api/src/shared/entity/base_entity.dart';
import 'package:auto_mapper/auto_mapper_annotation.dart';

@AutoMap(target: MembroDto)
class MembroEntity extends BaseEntity {
  final String nomeMembro;
  final String? emailMembro;
  final String? contatoMembro;
  final String? enderecoMembro;
  final StatusAssociado statusAssociado;
  final int quantidadePresenca;

  MembroEntity({
    required super.id,
    required super.criadoEm,
    required super.atualizadoEm,
    required super.deletadoEm,
    required this.nomeMembro,
    required this.emailMembro,
    required this.contatoMembro,
    required this.enderecoMembro,
    required this.statusAssociado,
    required this.quantidadePresenca,
  });

  MembroEntity copyWith({
    String? nomeMembro,
    String? emailMembro,
    String? contatoMembro,
    String? enderecoMembro,
    StatusAssociado? statusAssociado,
    int? quantidadePresenca,
    DateTime? deletadoEm,
  }) {
    return MembroEntity(
      id: id,
      criadoEm: criadoEm,
      atualizadoEm: DateTime.now(),
      deletadoEm: deletadoEm ?? this.deletadoEm,
      nomeMembro: nomeMembro ?? this.nomeMembro,
      emailMembro: emailMembro ?? this.emailMembro,
      contatoMembro: contatoMembro ?? this.contatoMembro,
      enderecoMembro: enderecoMembro ?? this.enderecoMembro,
      statusAssociado: statusAssociado ?? this.statusAssociado,
      quantidadePresenca: quantidadePresenca ?? this.quantidadePresenca,
    );
  }
}
